### DISCUSSION TITLE
Component Name/ Function in LWC Log Entry are null.

### DISCUSSION THREAD

**Discussion by robferro** [jongpie/NebulaLogger#657] [Q&A] [answered]
Hello!

I'm testing out the unmanaged package and noticed that when an LWC logs an entry, the Component Name and Function are null. I can find the LWC name buried in the stack trace, but I was wondering if there was a way to populate these fields?

I'm using the `import { createLogger } from 'c/logger';` method and below is the snippet that creates the log:

`this.logger.error(errMessage);
 this.logger.saveLog().then(() => {
       console.log('done with async save');
  });`

![image](https://github.com/jongpie/NebulaLogger/assets/71719385/2cd18df4-5897-446b-9308-4441722a34b1)

**Comment by robferro** [comment]
I tried using this method based on the recipe:

` const entry = this.logger.error(this.message).setError(someError).addTag('lwc logging demo');`

But anytime I add the setError method, the dev console throws an error:

![image](https://github.com/jongpie/NebulaLogger/assets/71719385/c85aec85-311d-4470-9f63-849e1961e57b)

**Reply by jongpie** [reply]
Hi @robferro - which version of Nebula Logger are you using?

**Reply by robferro** [reply]
@jongpie Version Number 4.13.1

**Reply by jongpie** [reply]
Perfect, thanks! Those fields are supposed to be automatically set, so I'll need to do some analysis to see what's going on. I'll let you know what I find out ASAP.

**Reply by robferro** [reply]
@jongpie Hey! Did you have a chance to analyze it to see what's happening? Thanks!

**Reply by jongpie** [reply]
Hey @robferro , sorry for not following up sooner! I've tried to reproduce the issue, but so far, everything seems to be working for me. I've been testing with [this demo LWC](https://github.com/jongpie/NebulaLogger/blob/v4.13.7/nebula-logger/recipes/lwc/loggerLWCDemo/loggerLWCDemo.js), and it seems to always populate the component name & function name (see screenshot below).

![image](https://github.com/jongpie/NebulaLogger/assets/1267157/bb437c0b-3e31-43f5-982a-c4d07f3ecdaa)

- Could you take a look at the demo LWC, and see if there's anything different about how you're logging in your LWC? 
- Could you also check your org's Session Settings (under Setup), and see what options are selected for Lighting Locker API Version and Lightning Web Security (screenshot below)? These session settings have caused problems in the past, so they might be the source of the problem.
  
  ![image](https://github.com/jongpie/NebulaLogger/assets/1267157/07426bd4-1faa-4d29-8e4a-beb32b70613e)

**Reply by robferro** [reply]
@jongpie No dice on that setting unfortunately. I copied the exact same code from the `connectedCallback` function in your demo. And I see the below log entry:
![image](https://github.com/jongpie/NebulaLogger/assets/71719385/bafd8a0b-bbde-4388-8139-bfaeb6dd7d3f)

This is the code:

````
import { LightningElement, api } from 'lwc';
import { createRecord } from "lightning/uiRecordApi";
import { ShowToastEvent } from "lightning/platformShowToastEvent";

import ACCOUNT_OBJECT from "@salesforce/schema/Account";
import NAME_FIELD from "@salesforce/schema/Account.Name";

import { createLogger } from 'c/logger';

export default class TestLWC extends LightningElement {
    @api
    logger;

    name = "";

    async connectedCallback() {
      this.logger = await createLogger();
      try {
        this.logger.error('test error entry');
        this.logger.warn('test warn entry');
        this.logger.info('test info entry');
        this.logger.debug('test debug entry');
        this.logger.fine('test fine entry');
        this.logger.finer('test finer entry');
        this.logger.finest('test finest entry');
        throw new Error('A bad thing happened here');
    } catch (error) {
        this.logger.error('>>> connectedCallback error: ' + error.message).setError(error);
        this.logger.saveLog().then(() => {
            console.log('done with async save');
        });
    }
    }

**Reply by robferro** [reply]
I noticed that I can see the component name in the stack trace. Is there something strange in there that could be messing up the logic to extract the component name?
![image](https://github.com/jongpie/NebulaLogger/assets/71719385/bd1e6e96-4d4c-4ad2-9a41-6626a36d6353)

**Reply by jongpie** [reply]
@robferro there is definitely something strange in the stack trace that you're seeing - although it has the component name, it seems to be missing the function name, which is causing problems with parsing the data. When testing this in my own scratch orgs, the stack traces have a line that contain the function name and the component name, like:

```javascript
// Component name: c/loggerLWCDemo, function name: renderedCallback
renderedCallback@modules/c/loggerLWCDemo.js:225:26
```

```javascript
// Component name: c/loggerLWCDemo, function name: logInfoExample
logInfoExample@modules/c/loggerLWCDemo.js:3:5759
```

These lines (example shown in the screenshot below) are parsed to determine the component name & function - but since the stack traces in your org don't seem to have the same format, Nebula Logger isn't able to parse the data. Looking at the stack trace you provided, it doesn't seem to contain the function name anywhere, so I'm not sure how to resolve this problem.

![image](https://github.com/jongpie/NebulaLogger/assets/1267157/77d1a153-5542-4eb3-b48e-75b77dd9e7cd)

**Reply by robferro** [reply]
I'm not sure either... We'll see if we can find a pattern for the root cause once we go live with the framework. Thanks for taking a look!

**Reply by jongpie** [reply]
That sounds great, thanks @robferro. My guess is it's still somehow related to an org setting somewhere under `Setup` (even though those 2 session settings didn't seem to explain it). If you find any pattern in your orgs, or notice a difference in the format of the stack traces, let me know. I'll also try to continue to reproduce the issue on my side.

**Reply by jongpie** [reply]
@robferro I realized I haven't posted an updated on this discussion in a while, but there have been some good improvements made with JS stack trace parsing over the last few months:
- In [release `v4.13.17`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.13.17), I completely rewrote how JS stack trace parsing works - it's now much better overall, and has much better cross-browser support
- Issue #728 was 1 issue not resolved by `v4.13.17` - JS stack traces in `async` functions still don't have fully accurate function names. This is a challenge in JavaScript in general (not an issue specific to Nebula Logger) because some browsers simply don't provide enough details in stack traces generated in `async` functions
- PR #775 will help avoid issue #728 - but not directly fix it - by eliminating the need to use `await` when importing Nebula Logger's component
  - For example, in [your comment above](https://github.com/jongpie/NebulaLogger/discussions/657#discussioncomment-9159517) the `connectedCallback()` was `async` because `createLogger()` requires it - here's a simplified version of it:
  
      ```javascript
        import { LightningElement } from 'lwc';
        import { createLogger } from 'c/logger';

        export default class TestLWC extends LightningElement {
            logger;

            async connectedCallback() {
                this.logger = await createLogger();
                this.logger.info('test info entry, after using async & await');
                this.logger.saveLog();
            }
        }
      ```
  - Once PR #775 is merged (hopefully this week), `async` won't be needed, and the resulting stack traces will be more accurate
  
    ```javascript
        import { LightningElement } from 'lwc';
        import { getLogger } from 'c/logger';

        export default class TestLWC extends LightningElement {
            logger = getLogger();

            connectedCallback() {
                this.logger.info('test info entry, no await required');
                this.logger.saveLog();
            }
        }
    ```

**Reply by pcc-ferror** [reply]
@jongpie Glad to hear! I'll subscribe to that PR to stay updated. Cheers!

**Reply by jongpie** [reply, marked as answer]
Alright, another update - I've just published [release `v4.14.13`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.14.13), which I think finishes addressing all of the open/known issues with LWC logging. There are still some known limitations with logging in `async` functions (mentioned in the release notes), but I'm not aware of any other issues.

I'm marking this discussion as resolved for now, but we can create new issues/discussions if something else comes up.
